Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs

```diff
@@ -0,0 +1,318 @@
+using Microsoft.MixedReality.Toolkit.Internal.EventDatum.Teleport;
+using Microsoft.MixedReality.Toolkit.Internal.Utilities.Physics;
+using System;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.SDK.UX.Pointers
+{
+    public class TeleportPointer : LinePointer
+    {
+        [SerializeField]
+        private MixedRealityInputAction teleportAction = MixedRealityInputAction.None;
+
+        [SerializeField]
+        [Tooltip("The angle from the pointer's forward position that will activate the teleport.")]
+        private float activationAngle = 45f;
+
+        [SerializeField]
+        [Tooltip("If Pressing 'forward' on the thumbstick gives us an angle that doesn't quite feel like the forward direction, we apply this offset to make navigation feel more natural")]
+        private float angleOffset = 0f;
+
+        [SerializeField]
+        private float upDirectionThreshold = 0.2f;
+
+        [SerializeField]
+        protected Gradient LineColorHotSpot = new Gradient();
+
+        [SerializeField]
+        [Tooltip("Layers that are considered 'valid' for navigation")]
+        protected LayerMask ValidLayers = 1 << 0; // Default
+
+        [SerializeField]
+        [Tooltip("Layers that are considered 'invalid' for navigation")]
+        protected LayerMask InvalidLayers = 1 << 2; // Ignore raycast
+
+        [SerializeField]
+        private float inputThreshold = 0.5f;
+
+        private Vector2 currentInputPosition = Vector2.zero;
+
+        private bool teleportEnabled = false;
+
+        private bool canTeleport = false;
+
+        /// <summary>
+        /// The result from the last raycast.
+        /// </summary>
+        public TeleportSurfaceResult TeleportSurfaceResult { get; private set; } = TeleportSurfaceResult.None;
+
+        protected Gradient GetLineGradient(TeleportSurfaceResult targetResult)
+        {
+            switch (targetResult)
+            {
+                case TeleportSurfaceResult.None:
+                    return LineColorNoTarget;
+                case TeleportSurfaceResult.Valid:
+                    return LineColorValid;
+                case TeleportSurfaceResult.Invalid:
+                    return LineColorInvalid;
+                case TeleportSurfaceResult.HotSpot:
+                    return LineColorHotSpot;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(targetResult), targetResult, null);
+            }
+        }
+
+        protected override void OnEnable()
+        {
+            base.OnEnable();
+            BaseCursor?.SetVisibility(false);
+        }
+
+        #region IMixedRealityPointer Implementation
+
+        /// <inheritdoc />
+        public override bool IsInteractionEnabled => !IsTeleportRequestActive && teleportEnabled;
+
+        /// <inheritdoc />
+        public override float PointerOrientation
+        {
+            get
+            {
+                if (
+                    TeleportHotSpot != null &&
```

Comment:
nit: can you move this up to the same line as the (?

Reply:
yes, I'm not sure how that got passed me. lol